odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+            raise TypeError('second temporary {!r} not in the domain '
+                            '{!r} of the functional.'
+                            ''.format(tmp2, self._left.domain))
+        self._tmp2 = tmp2
+
+    @property
+    def gradient(self):
+        """Gradient of the compositon according to the chain rule."""
+
+        func = self.left
+        op = self.right
+
+        class CompositGradient(Operator):
+            """Gradient of the compositon according to the chain rule."""
+            def __init__(self):
+                """Initialize a new instance."""
+                super().__init__(func.domain, func.domain, linear=False)
+                self._func = func
+                self._op = op
+
+            def _call(self, x):
+                """Applies the gradient operator to the given point.
+
+                Parameters
+                ----------
+                x : `LinearSpaceVector`
+                    Element in the domain of the functional to which the
+                    gradient operator is applied. The element must have a
+                    non-zero norm
+
+                Returns
+                -------
+                `self(x)` : `LinearSpaceVector`
+                    Evaluation of the gradient operator. An element in the
+                    domain of the functional.
+                """
+                return self._op.derivative(x).adjoint(
+                    self._func.gradient(self._op(x)))
+
+        return CompositGradient()
+
+
+class FunctionalRightVectorMult(Functional, OperatorRightVectorMult):
+
+    """Expression type for the functional right vector multiplication.
+
+    Given a functional ``func`` and a vector ``y`` in the domain of ``func``,
+    this corresponds to the functional
+
+        ``(func * y)(x) = func(y*x)``.
+    """
+
+    def __init__(self, func, vector):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The domain of ``func`` must be a ``vector.space``.
+        vector : `LinearSpaceVector` in ``func.domain``
+            The vector to multiply by.
+        """
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        OperatorRightVectorMult.__init__(self, operator=func, vector=vector)
+
+        Functional.__init__(self, domain=func.domain)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.vector * self.operator.gradient * self.vector
+
+    # TODO: can this be computed?
```

Comment:
Probably not. The multiplication operator does not fulfill the "almost unitary" property that allows computation of the proximal of the composition.